### PR TITLE
Implement copy when size is not known

### PR DESCRIPTION
**Pull request by haichangsi**
Implement a collective copy without providing the size of the source in the destination.

### CODE REVIEW

File: test/gtest/cpu/algorithms.cpp [hunk 1]
```diff
@@ -36,6 +36,15 @@ void check_copy(std::size_t n, std::size_t b, std::size_t e) {
...
     std::copy(dv_in.begin() + b, dv_in.begin() + e, dv3.begin() + b);
   }
+  if (comm_rank == 0) {
+    lib::copy(0, v_in.begin() + b, v_in.begin() + e, dv8.begin() + b);
+  } else {
+    lib::copy(0, nullptr, nullptr, dv8.begin() + b);
+  }
+
+  lib::copy(0, comm_rank == 0 ? &*(v_in.begin() + b) : nullptr,
+            comm_rank == 0 ? &*(v_in.begin() + e) : nullptr, dv9.begin() + b);
```
Comment (KFilipek) [inline test/gtest/cpu/algorithms.cpp:46]: Unnecessary obfuscation:
```cpp
 lib::copy(0, comm_rank == 0 ? &*(v_in.begin() + b) : nullptr,
            comm_rank == 0 ? &*(v_in.begin() + e) : nullptr, dv9.begin() + b);
```
=>
```cpp
if (comm_rank == 0) {
    lib::copy(0, &*(v_in.begin() + b), &*(v_in.begin() + e), dv9.begin() + b);
} else {
    lib::copy(0, nullptr, nullptr, dv9.begin() + b);
}
```

Generally, the whole function should be rewritten, for now, it's not human-readable.

**Review by KFilipek — CHANGES_REQUESTED** [review]
Please split those changes into two commits:
1. that adds the `bcast()` function
2. rest of the implementation
The background of splitting commits into atomic/functional parts is that you can revert some changes without losing interesting features.

File: include/dr/details/communicator.hpp [hunk 2]
```diff
@@ -80,6 +80,10 @@ class communicator {
 
   void barrier() const { MPI_Barrier(mpi_comm_); }
 
+  template <typename T> void bcast(T *src, size_t count, int root) const {
+    MPI_Bcast(src, count * sizeof(T), MPI_CHAR, root, mpi_comm_);
```
[the remark this answers is not in the document]
Comment (rscohn2) [reply include/dr/details/communicator.hpp:84]: I didn't know about MPI_BYTE. Can you fix the other instances in the same file?

File: include/dr/algorithms/algorithms.hpp [hunk 3]
```diff
@@ -199,7 +193,24 @@ void copy(int root, std::nullptr_t, std::size_t size,
                 scatter_data_dst_size(result, size, comm.rank()), root);
 }
 
-///// Collective copy from local range to distributed iterator
+/// Collective copy from local begin/end to distributed, which is called on
+/// the non-source rank without providing the size of the source (if I is the
+/// same as std::nullptr_t). Collective copy from local begin/end to
```
Comment (rscohn2) [inline include/dr/algorithms/algorithms.hpp:198]:
```suggestion
/// same as `std::nullptr_t`). Collective copy from local begin/end to
```
Delimiters excludes it from spell check and will also format it appropriately.

File: include/dr/algorithms/algorithms.hpp [hunk 4]
```diff
@@ -199,7 +193,24 @@ void copy(int root, std::nullptr_t, std::size_t size,
...
+/// distributed, which is called on the source rank without providing the size
+/// of the source in the destination (if I is not the same as std::nullptr_t)
```
Comment (rscohn2) [inline include/dr/algorithms/algorithms.hpp:200]:
```suggestion
/// of the source in the destination (if `I` is not the same as `std::nullptr_t`)
```

File: test/gtest/cpu/algorithms.cpp (hunk 1 above)
Comment (lslusarczyk) [reply test/gtest/cpu/algorithms.cpp:46]: This in on purpose. To have "?:" which casts third expression into type of second one.

File: include/dr/details/communicator.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (lslusarczyk) [reply include/dr/details/communicator.hpp]: applied

File: test/gtest/cpu/algorithms.cpp (hunk 1 above)
Comment (lslusarczyk) [reply test/gtest/cpu/algorithms.cpp:46]: not applied, "?:" is on purpose

**Comment by lslusarczyk** [comment]
Code of this PR is reported in other PR. https://github.com/oneapi-src/distributed-ranges/pull/11
Closing this one.

**Pull request closed without merging**